module/zfs/dmu_object.c

```diff
@@ -61,6 +61,7 @@ dmu_object_alloc_dnsize(objset_t *os, dmu_object_type_t ot, int blocksize,
 	boolean_t restarted = B_FALSE;
 	uint64_t *cpuobj = NULL;
 	int dnodes_per_chunk = 1 << dmu_object_alloc_chunk_shift;
+	int error;
 
 	kpreempt_disable();
 	cpuobj = &os->os_obj_next_percpu[CPU_SEQID %
```

Comment:
Thanks for the clarification. Regardless, the atomic update could use `atomic_cas_64()` to only update the variable if the value has not been changed by another thread in the meantime. This might make things less fragile.
Performance wise, cache lines of `os_obj_next_percpu` will be bounced a lot under heavy multithreaded load. This could be easily resolved by adding sufficient padding for each per-cpu value.